### PR TITLE
use macros for `projectRuffExecutablePath`

### PR DESCRIPTION
**Pull request by KotlinIsland**
- Attempting to address #343

The code in here works, but is really bad. I couldn't figure out how to make macros work properly.

Currently the plugin always saves the full path to the config file. This makes it impossible to commit project config. In this change we collapse the venv dir before it is saved.

### CODE REVIEW

**Comment by koxudaxi** [comment]
@KotlinIsland 
I'm sorry for my late reply.
I tested the your PR in my windows OS.
image


The ruff binary file name includes the `.exe` suffix.
Do you expect the behavior? 
Should you use the special variable for ruff binary, like `$ruff$` to inject the actual binary file name for each OS?

**Comment by KotlinIsland** [comment]
Ahh, good point, I didn't anticipate this complication. I see two possibilities:
1. We point to the directory containing the executable.
2. We automatically handle adding/removing the `.exe` as needed.

**Comment by koxudaxi** [comment]
> We point to the directory containing the executable, instead of the executable itself.

It is certainly easier to save to file this way. However, for clarity regarding the display on the settings screen, it might be better to include up to .exe.

**Comment by KotlinIsland** [comment]
> for clarity

I completely agree.

**Comment by koxudaxi** [comment]
> Additionally, this PR contains only a rough draft, I couldn't figure out how to make macro fields work properly.
OK, I will check it today.

File: src/com/koxudaxi/ruff/Ruff.kt [hunk 1]
```diff
@@ -59,6 +59,7 @@ val RUFF_LSP_COMMAND = when {
     else -> "ruff-lsp"
 }
 const val WSL_RUFF_LSP_COMMAND = "ruff-lsp"
+const val PY_INTERPRETER_DIRECTORY_MACRO_NAME = "PY_INTERPRETER_DIRECTORY"
```
Comment (koxudaxi) [inline src/com/koxudaxi/ruff/Ruff.kt:62]: Should we use Screaming snake case?

File: src/com/koxudaxi/ruff/RuffPackageManagerListener.kt [hunk 2]
```diff
@@ -9,15 +9,27 @@ import com.jetbrains.python.packaging.PyPackageManager
...
             @Suppress("UnstableApiUsage")
             lspServerManager.stopAndRestartIfNeeded(RuffLspServerSupportProvider::class.java)
         }
     }
+
+    private fun updateMacro(sdk: Sdk) {
+        sdk.homeDirectory?.parent?.presentableUrl?.let {
+            ruffMacroService.addMacroExpand(PY_INTERPRETER_DIRECTORY_MACRO_NAME, it)
+        }
+    }
```
Comment (koxudaxi) [inline src/com/koxudaxi/ruff/RuffPackageManagerListener.kt:34]: The timing of the sdk path change is
I want to fire this event there if and when the sdk path changes.

File: src/com/koxudaxi/ruff/RuffConfigService.kt [hunk 3]
```diff
@@ -1,13 +1,11 @@
...
-import com.intellij.openapi.components.Storage
+import com.intellij.openapi.components.*
 import com.intellij.openapi.project.Project
 import com.intellij.util.xmlb.XmlSerializerUtil
 import org.jetbrains.annotations.SystemDependent
 
+
 @State(name = "RuffConfigService", storages = [Storage("ruff.xml")])
 @Service(Service.Level.PROJECT)
 class RuffConfigService : PersistentStateComponent<RuffConfigService> {
```
Comment (koxudaxi) [inline src/com/koxudaxi/ruff/RuffConfigService.kt:11]: If I add the `project: Project` in the argument, the service raises a null point exception.
I had no choice but to create RuffMacroService.

**Review by koxudaxi — COMMENTED** [review]
@KotlinIsland 
I made some modifications.
Add a logic to save to file also when path is overide if necessary, because the timing written to config is only when it should be applied in config panel.

I don't know why, but RuffConfigService can't take project: project in its constructor, so the code became quite complicated. I'd like to make it a little more transparent.

File: src/com/koxudaxi/ruff/Ruff.kt (hunk 1 above)
Comment (KotlinIsland) [reply src/com/koxudaxi/ruff/Ruff.kt:62]: Yes: https://kotlinlang.org/docs/coding-conventions.html#property-names

**Comment by koxudaxi** [comment]
@KotlinIsland 

I did some more research.
Here is a code that might be helpful.
https://github.com/JetBrains/intellij-community/blob/94282b7d391a5ed702f01a31549b4893c62974f1/python/src/com/jetbrains/python/sdk/PySdkSettings.kt#L52-L81

I think that before writing the Path to the config file, we can set the macro as inject and expand it again if we want to refer to it.
However, I think there needs to be some caching mechanism to avoid the overhead of executing the code here every time it is referenced.

**Comment by koxudaxi** [comment]
@KotlinIsland 
I thought about it some more, but I felt that if We don't need to use a macro to select the bin of the project, if `ruff plugin` gets the `ruff` or `ruff.exe`.
In other words, what if the user can choose to either use the project's venv bin or a custom executable?
I think the custom executables should not be shared with another developer in git.

**Comment by koxudaxi** [comment]
@KotlinIsland 
What do you think?

**Comment by KotlinIsland** [comment]
Looks great!

**Pull request closed without merging**